### ISSUE TITLE
Need Efficient ClassBalancedPolicy 

### ISSUE THREAD

**Issue by Mattdl** [ContinualAI/avalanche#653]
Coming from #650 , the ClassBalancedPolicy (and probably ExperienceBalancedPolicy) seems to cause a significant slowdown for the data incremental setting. In the Class-incremental setting it doesn't have much effect, but the data incremental with batch-wise processing (each experience is a batch) needs a more efficient implementation. It remains to be figured out which part is causing the slowdown (a hunch is that it could be the concat of too many AvalancheDatasets).
Initially, storing a new batch takes about 0.001 seconds, and after about 100 experiences (batches) takes about 0.1 seconds.
Below is a working example:

    from __future__ import absolute_import
    from __future__ import division
    from __future__ import print_function
    
    import torch
    import torch.optim.lr_scheduler
    
    from avalanche.benchmarks import SplitMNIST
    from avalanche.models import SimpleMLP
    from avalanche.training.strategies import Naive
    from avalanche.training.plugins import ReplayPlugin
    from avalanche.evaluation.metrics import StreamForgetting, \
        accuracy_metrics, loss_metrics
    from avalanche.logging import TextLogger
    from avalanche.training.plugins import EvaluationPlugin
    from avalanche.benchmarks.generators.benchmark_generators import \
        data_incremental_benchmark
    
    
    from avalanche.training.plugins import ClassBalancedStoragePolicy
    
    def main(args):
        # --- DEFAULT PARAMS ONLINE DATA INCREMENTAL LEARNING
        nb_tasks = 5  # Can still design the data stream based on tasks
        batch_size = 10  # Learning agent only has small amount of data available
        epochs = 1  # How many times to process each mini-batch
        return_task_id = True  # Data incremental (task-agnostic/task-free)
    
        # --- CONFIG
        device = torch.device(
            f"cuda:{args.cuda}" if torch.cuda.is_available() and args.cuda >= 0
            else "cpu")
        # ---------
    
        # --- SCENARIO CREATION
        task_scenario = SplitMNIST(nb_tasks, return_task_id=return_task_id,
                                   fixed_class_order=[i for i in range(10)])
    
        # Make data incremental (one batch = one experience)
        scenario = data_incremental_benchmark(task_scenario,
                                              experience_size=batch_size)
        print(f"{scenario.n_experiences} batches in online data incremental setup.")
        # 6002 batches for SplitMNIST with batch size 10
    
        # MODEL CREATION
        model = SimpleMLP(num_classes=args.featsize,
                          hidden_size=400, hidden_layers=2)
    
        # choose some metrics and evaluation method
        interactive_logger = TextLogger()
        eval_plugin = EvaluationPlugin(
            accuracy_metrics(experience=False, stream=True),
            loss_metrics(experience=False, stream=True),
            StreamForgetting(),
            loggers=[interactive_logger])
        sp = ClassBalancedStoragePolicy(ext_mem={},mem_size=2000)
    
        cl_strategy = Naive(model, torch.optim.SGD(model.parameters(), lr=0.01),
                            torch.nn.CrossEntropyLoss(),
                            train_mb_size=batch_size, train_epochs=epochs,
                            eval_mb_size=100, device=device,
                            plugins=[ReplayPlugin(mem_size=2000,
                                                  storage_policy=sp)],
                            evaluator=eval_plugin
                            )
    
        # TRAINING LOOP
        results = []
        for experience in scenario.train_stream:
            print("Start of experience ", experience.current_experience)
            cl_strategy.train(experience)
        print('Computing accuracy on the whole test set')
        results.append(cl_strategy.eval(scenario.test_stream))
    
    
    if __name__ == '__main__':
        import argparse
        parser = argparse.ArgumentParser()
        parser.add_argument('--cuda', type=int, default=0,
                            help='Select zero-indexed cuda device. -1 to use CPU.')
        parser.add_argument('--featsize', type=int, default=100,
                            help='Feature size for the embedding.')
        args = parser.parse_args()
        main(args)

**Comment by AntonioCarta** [comment]
from a quick profiling, you seem correct. Most of the time is spent inside `__getitem__`.

![immagine](https://user-images.githubusercontent.com/8118624/120987764-342b6d00-c77e-11eb-85a3-3d1ed9f06063.png)

@lrzpellegrini can we do something about it? My impression is that right now `AvalancheDatasets` are quite similar to pytorch's Datasets, which are not designed to support large numbers of concatenations. We should have a "fast path" for concatenations/subsampling, where if two datasets come from the same source, then it should be possible to merge them without creating a `ConcatDataset`.

In the worse case, if the previous solution is not possible, we should at least resort to something like a TensorDatasets, which can be easily concatenated and subsampled.

**Comment by lrzpellegrini** [comment]
Using Subset/ConcatDatasets allows for data to be kept on disk instead of being stored in memory. The fact that `_get_single_item` takes so much time suggests that the instance loading and preprocessing procedures may be too slow when replay is used in an intensive way, so it would be nice to expose an option to store data in memory as tensors instead of using this lazy loading mechanism (like a bool flag in the replay plugin). In this case we have to make sure the equivalence holds from a "transformations" perspective. This would be a nice addition per se.

Then, as @AntonioCarta suggested, some kind of "fast path" is totally doable, but this may require some optimizations on how indices/lists of datasets are kept internally in AvalancheSubset/ConcatDataset. These wrappers keep a strict ordering of instances (same of the PyTorch counterpart), but we may also consider adding an "ignore_order" flag to allow for this kind of optimizations. I have a bad feeling that continuously subsetting and concatenating (which is the standard procedure in ReplayPlugin) creates some "fragmentation" at the ConcatDataset level and this is why some additional overhead is being added.

However, before venturing in AvalancheDataset internals, I'd start from the in-memory storage of replay instances and then do another round of tests to check if that was the actual problem.

**Comment by AntonioCarta** [comment]
> The fact that _get_single_item takes so much time suggests that the instance loading and preprocessing procedures may be too slow when replay is used in an intensive way, so it would be nice to expose an option to store data in memory as tensors instead of using this lazy loading mechanism 

The loading time is constantly growing, so it seems like these operations become more expensive over time, which is unexpected.

> These wrappers keep a strict ordering of instances (same of the PyTorch counterpart), but we may also consider adding an "ignore_order" flag to allow for this kind of optimizations

Do we have a use case that needs strict ordering? I can't imagine anything.

**Comment by lrzpellegrini** [comment]
> Do we have a use case that needs strict ordering? I can't imagine anything.

For the replay plugin the ordering in fact doesn't make much sense to me, too. The ordering thing is related to how the datasets behave. Basically, concat dataset keeps the order of concatenated datasets "as-is" and the same applies to "subset", which keeps the list of indices without reordering them. This is the expected behavior when using these classes. However, adding an "ignore_order" flag may allow for more freedom when in the context of replay.
